test/SadTest.cc

```diff
@@ -0,0 +1,500 @@
+        } break;
+        case Unalign: {
+            for (int i = 0; i < MAX_BLOCK_SIZE; i++) {
+                source_data_[i] = rnd->random();
+                ref_1_data_[i] = rnd->random();
+                ref_2_data_[i] = rnd->random();
+            }
+            source_stride_ -= 1;
+            ref_1_stride_ -= 1;
+            ref_2_stride_ -= 1;
+        } break;
+        default: break;
+        }
+        delete rnd;
+    };
+
+    uint32_t reference_sad() {
+        unsigned int sad = 0;
+        for (int h = 0; h < height_; ++h) {
+            for (int w = 0; w < width_; ++w) {
+                sad += abs(source_data_[h * source_stride_ + w] -
+                           ref_1_data_[h * ref_1_stride_ + w]);
+            }
+        }
+        return sad;
+    }
+
+    unsigned int reference_sad_avg() {
+        unsigned int sad = 0;
+        for (int h = 0; h < height_; ++h) {
+            for (int w = 0; w < width_; ++w) {
+                const int tmp = ref_2_data_[h * ref_2_stride_ + w] +
+                                ref_1_data_[h * ref_1_stride_ + w];
+                const uint8_t comp_pred = ROUND_POWER_OF_TWO(tmp, 1);
+                sad += abs(source_data_[h * source_stride_ + w] - comp_pred);
+            }
+        }
+        return sad;
+    }
+
+  protected:
+    int width_, height_;
+    int source_stride_;
+    int ref_1_stride_;
+    int ref_2_stride_;
+
+    DECLARE_ALIGNED(32, uint8_t, source_data_[MAX_BLOCK_SIZE]);
+    DECLARE_ALIGNED(32, uint8_t, ref_1_data_[MAX_BLOCK_SIZE]);
+    DECLARE_ALIGNED(32, uint8_t, ref_2_data_[MAX_BLOCK_SIZE]);
+};
+/**
+ * @brief Unit test for SAD sub smaple functions include:
+ *  - NxMSadKernelSubSampled_funcPtrArray
+ *  - NxMSadKernelSubSampled_funcPtrArray
+ *
+ * Test strategy:
+ *  This test case combine different wight(4-64) x height(4-64), different test
+ * vecotr pattern(MaxRef, MaxSrc, Random, Unalign) to generate test vector. And
+ * select correct function from funcPtrArray, run func with test vector, compare
+ * result between reference function, non_avx2 function and avx2 function.
+ *
+ *
+ * Expect result:
+ *  Results come from reference functon, non_avx2 function and avx2 funtion are
+ * equal.
+ *
+ * Test coverage:
+ *  All functions inside NxMSadKernelSubSampled_funcPtrArray and
+ * NxMSadKernelSubSampled_funcPtrArray.
+ *
+ * Test cases:
+ *  Width (4, 8, 16, 24, 32, 48, 64) x height 4, 8, 16, 24, 32, 48, 64)
+ *  Test vector pattern (MaxRef, MaxSrc, Random, Unalign).
+ *
+ */
+class SADTestSubSample : public ::testing::WithParamInterface<SadTestParam>,
+                         public SADTestBase {
+  public:
+    SADTestSubSample()
+        : SADTestBase(std::get<0>(TEST_GET_PARAM(1)),
+                      std::get<1>(TEST_GET_PARAM(1)), TEST_GET_PARAM(0)) {
+    }
+
+  protected:
+    void check_sad() {
+        EB_SADKERNELNxM_TYPE non_avx2_func = nullptr;
+        EB_SADKERNELNxM_TYPE avx2_func = nullptr;
+        switch (width_) {
```

Comment:
Maybe we can optimize this big switch case, such as:
non_avx2_func =NxMSadKernelSubSampled_funcPtrArray[ASM_NON_AVX2][width_>>3];
avx2_func = NxMSadKernelSubSampled_funcPtrArray[ASM_AVX2][width_>>3];